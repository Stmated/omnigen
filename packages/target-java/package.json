{
  "name": "@omnigen/target-java",
  "version": "0.0.0",
  "private": true,
  "license": "MIT",
  "type": "module",
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "exports": {
    "ast": "./src/ast/index.ts",
    "default": "./src/index.ts"
  },
  "scripts": {
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "dev": "tsc -w",
    "lint": "TIMING=1 eslint src/**/*.ts* --fix",
    "test": "vitest run",
    "test:watch": "vitest"
  },
  "dependencies": {
    "@omnigen/api": "workspace:*",
    "@omnigen/core": "workspace:*",
    "@omnigen/core-log": "workspace:*",
    "@omnigen/core-plugin": "workspace:*",
    "@omnigen/plugin": "workspace:*",
    "@omnigen/target-code": "workspace:*",
    "graphology": "^0.25.4",
    "graphology-dag": "0.4.1",
    "zod": "^3.25.76"
  },
  "devDependencies": {
    "@omnigen/config-typescript": "workspace:*",
    "@omnigen/eslint-config-custom": "workspace:*",
    "@omnigen/utils-test": "workspace:*",
    "java-parser": "^3.0.1",
    "typescript": "5.9.2"
  }
}
