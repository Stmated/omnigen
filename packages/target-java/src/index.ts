import {OMNI_GENERIC_FEATURES, TargetFeatures} from '@omnigen/core';

export * from './visit/index.js';
export * from './transform/index.js';
export * from './interpret/index.js';
export * from './options/index.js';
export * from './util/index.js';
export * from './render/index.js';
export * from './parse/index.js';
export * as Java from './ast/index.js';

export const JAVA_FEATURES: Readonly<TargetFeatures> = {
  ...OMNI_GENERIC_FEATURES,
  literalTypes: false,
  primitiveGenerics: false,
};
