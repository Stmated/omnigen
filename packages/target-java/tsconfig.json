{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@omnigen/config-typescript/build.json",
  "compilerOptions": {
    "rootDir": "./src",
    "baseUrl": "./src"
  }
}
