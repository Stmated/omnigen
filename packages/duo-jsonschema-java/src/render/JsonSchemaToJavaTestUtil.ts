import {RenderedCompilationUnit} from '@omnigen/core';
import {PluginManager} from '@omnigen/plugin';
import {JavaOptions, JavaPlugins} from '@omnigen/target-java';
import {LoggerFactory} from '@omnigen/core-log';
import {JsonSchemaPlugins} from '@omnigen/parser-jsonschema';
import {FileContext} from '@omnigen/core-plugin';

const logger = LoggerFactory.create(import.meta.url);

export class JsonSchemaToJavaTestUtil {

  public static async render(inPath: string, options?: JavaOptions): Promise<RenderedCompilationUnit[]> {

    const pm = new PluginManager({includeAuto: true});

    logger.info(`Running plugin manager with ${JsonSchemaPlugins.JsonSchemaPlugin.name} and ${JavaPlugins.JavaPluginInit.name}`);

    // pm.addPlugin(new JavaPlugin);

    const ctx: FileContext = {
      file: inPath,
      arguments: {
        ...options,
      },
      // target: 'java', // <-- implicit, because we have no other target plugin in this package
      // arguments: {
      //   ...all.modelTransformOptions,
      //   ...all.parserOptions,
      //   ...all.targetOptions,
      //   ...all.packageOptions,
      //   ...all.javaOptions,
      //   ...all.arguments,
      // },
    };

    const result = await pm.execute({
      ctx: ctx,
      stopAt: JavaPlugins.JavaRendererCtxOut,
    });

    return result.result.ctx.compilationUnits;
  }
}
