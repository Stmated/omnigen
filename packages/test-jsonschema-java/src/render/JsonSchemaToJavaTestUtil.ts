import {RenderedCompilationUnit, TargetOptions} from '@omnigen/api';
import {PluginManager} from '@omnigen/plugin';
import {JavaOptions, JavaPlugins, SerializationLibrary} from '@omnigen/target-java';
import {LoggerFactory} from '@omnigen/core-log';
import {JsonSchemaPlugins} from '@omnigen/parser-jsonschema';
import {FileContext} from '@omnigen/core-plugin';
import {DEFAULT_SPECIFIC_TEST_TARGET_OPTIONS} from '@omnigen/utils-test';

const logger = LoggerFactory.create(import.meta.url);

/**
 * TODO: This and the same object from openrpc java test package should be in one place
 */
export const DEFAULT_SPECIFIC_TEST_JAVA_OPTIONS: Partial<JavaOptions> = {
  serializationLibrary: SerializationLibrary.POJO,
  beanValidation: false,
};

export class JsonSchemaToJavaTestUtil {

  public static async render(inPath: string, options?: Partial<JavaOptions & TargetOptions>): Promise<RenderedCompilationUnit[]> {

    const pm = new PluginManager({includeAuto: true});

    logger.info(`Running plugin manager with ${JsonSchemaPlugins.JsonSchemaPlugin.name} and ${JavaPlugins.JavaPluginInit.name}`);

    const ctx: FileContext = {
      file: inPath,
      defaults: {
      },
      arguments: {
        ...DEFAULT_SPECIFIC_TEST_JAVA_OPTIONS,
        ...DEFAULT_SPECIFIC_TEST_TARGET_OPTIONS,
        ...options,
      },
    };

    const result = await pm.execute({
      ctx: ctx,
      stopAt: JavaPlugins.JavaRendererCtxOut,
    });

    return result.result.ctx.compilationUnits;
  }
}
