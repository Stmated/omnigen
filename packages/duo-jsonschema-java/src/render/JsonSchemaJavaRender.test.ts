import {describe, expect, test, vi} from 'vitest';
import {JsonSchemaToJavaTestUtil} from './JsonSchemaToJavaTestUtil.ts';
import {ZodJavaOptions} from '@omnigen/target-java';

describe('jsonschema-java-render', () => {

  test('string_union', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render('./examples/string_union.json');
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu).toHaveLength(1);
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('x-enum-varnames', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render('../parser-jsonschema/examples/keep_x_enum_varnames.json');
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('decorated_types', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render(
      '../parser-jsonschema/examples/decorated_types.json',
      {...ZodJavaOptions.parse({}), commentsOnFields: false, commentsOnGetters: true, serializationLibrary: 'POJO'},
    );
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('enum_string_composition', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render('./examples/enum_string_composition.json', {
      compressSoloReferencedTypes: false,
      serializationLibrary: 'POJO',
    });
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('interface_order', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render('./examples/interface_order.json', {
      compressSoloReferencedTypes: false,
      // serializationLibrary: 'POJO',
    });
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  /**
   * Current result is not the best, since it creates an unused `StringBoolean` when inheriting `AccountNotificationDataVerified` is used instead.
   */
  test('inline_boolean', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render('./examples/inline_boolean.json', {
      serializationLibrary: 'POJO',
      includeExampleCommentsMode: 'SKIP',
    });
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('class_union', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render('./examples/class_union.json', {
      compressSoloReferencedTypes: false,
      compressUnreferencedSubTypes: false,
      serializationLibrary: 'JACKSON',
      includeExampleCommentsMode: 'SKIP',
    });
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('enum_inheritance', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render('./examples/enum_inheritance.json');
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('one_of_same_type', async ({task}) => {

    // TODO: Fix so that it either prints it as a String, or creates a composition type -- right now it tries to print the *object* DateOfBirthOrOrganization
    //        Something is likely wrong with some asSuperType/asSubType after type guard changes

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const rendered = await JsonSchemaToJavaTestUtil.render('./examples/one_of_same_type.json');
    const fileContents = Map.groupBy(rendered, it => it.fileName);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, cu] of fileContents) {
      expect(cu[0].content).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });
});
