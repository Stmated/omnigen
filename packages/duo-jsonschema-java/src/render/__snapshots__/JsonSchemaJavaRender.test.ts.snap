// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsonschema-java-render > class_union 1`] = `
[
  "AbortData.java",
  "Data.java",
  "DefaultData.java",
  "DefaultDataInlineResult.java",
  "DescriptiveStringBoolean.java",
  "StringBoolean.java",
]
`;

exports[`jsonschema-java-render > decorated_types 1`] = `
[
  "Order.java",
  "Shop.java",
]
`;

exports[`jsonschema-java-render > enum_inheritance 1`] = `
[
  "InheritingKind1.java",
  "Kind1.java",
]
`;

exports[`jsonschema-java-render > enum_string_composition 1`] = `
[
  "DataEntry.java",
  "TransferState.java",
]
`;

exports[`jsonschema-java-render > inline_boolean 1`] = `
[
  "AccountNotificationData.java",
  "AccountNotificationDataVerified.java",
  "StringBoolean.java",
]
`;

exports[`jsonschema-java-render > interface_order 1`] = `
[
  "AbstractRequestData.java",
  "BalanceRequestData.java",
  "CancelChargeRequestData.java",
  "CancelChargeRequestDataBase.java",
  "ICancelChargeRequestDataBase.java",
]
`;

exports[`jsonschema-java-render > string_union 1`] = `
[
  "DateOfBirthOrOrganization.java",
]
`;

exports[`jsonschema-java-render > x-enum-varnames 1`] = `
[
  "SomeEnum.java",
]
`;
